Create Readme for x-adapter (#921)
empathyco/x: packages/x-adapter/README.md

# x-adapter

> **Empathy Search Adapter** is a library of utils to ease the communication with any API.

Author: It's the way I found it is named in docs: 
`Empathy Platform Search Adapter to communicate with the Empathy Platform Search API (or the Empathy Search Adapter to connect with the search API you are using).` [(in before you begin section)](https://docs.empathy.co/develop-empathy-platform/build-search-ui/web-archetype-development-guide.html#configuring-the-plugin)

Some features that it provides:

- Create an API request function based on a simple configuration.
- Allow to configure several endpoints by extending the initial configuration.
- Allow to configure the response/request mapping.
- Create mapping functions based on Schemas.

## Installation

```
npm i @empathyco/x-adapter
```

You can additionally install the
[@empathyco/x-types](https://github.com/empathyco/x/tree/main/packages/search-types) package and

> take the advantage of it in your project development.

Author: I'll do so. 
Based on this I see I was mistaken: I understood that they could be used for type safety, regardless of whether or not `x-components` was used. So thanks for the appointment 👀

## Configuration & Usage

> The `Adapter` instance is meant to be a configuration object for each `EndpointAdapter`, handling

Author: Hi @javieri-empathy, 🤔 the idea was to show that in the end, they would have an adapter object to export & use in their app. So should we add this part at the end? Or I'm mistaken/losing something??? 🙏

each data model that will need to be passed/mapped/transformed to meet your app’s environment and
the API requirements.

> ```tsx

Author: True! That happened by copy-pasting code blocks from the Notion app (it seems they codify Typescript using tsx 👀 ).

      // ... any options your API may need
    }
  },
  // we enhance you to declare your resquest/response mappers ouside this function to improve its legibility
  requestMapper: searchRequestMapper,
  responseMapper: searchResponseMapper
});
```

### Implement your own adapter using schemas

The `x-adapter` library includes a `schemaMapperFactory` function that creates a mapper function to
execute the schema you provide. You just need to create your request and response schemas to define
the structure of your data, and pass them as a parameter to the factory function.

```tsx
import { schemaMapperFactory } from '@empathyco/x-adapter';
import { SearchRequest } from '@empathyco/x-types';
import { searchRequestSchema } from '../../schemas/requests/search-request.schema';

// Import and use already existing types from your API source, or declare your own based on the response you get.
interface ApiRequestModel {
  q: string;
  limit?: number;
  offset?: number;
}
interface ApiResponseModel {
  items: T[];
  limit?: number;
  offset?: number;
  total: number;
}

// Maps both the request and the response to connect your model with API you are working with.

> const searchRequestSchema: Schema<SearchRequest, ApiRequestModel> = {

Reviewer: I miss some talk about the features that schema has, like using paths, functions, mapping arrays, type safety...

  q: 'query',
  limit: 'rows',
  offset: 'start'
};
const searchResponseSchema: Schema<ApiResponseModel, SearchResponse> = {
  results: 'items',
  totalResults: 'items.total' // you can use paths to access to an inner property if needed
};

// The 'schemaMapperFactory' creates a mapper function for a given schema.
export const searchRequestMapper = schemaMapperFactory<SearchRequest, ApiRequestModel>(
  searchRequestSchema
);
export const searchResponseMapper = schemaMapperFactory<ApiResponseModel, SearchResponse>(
  searchResponseSchema
);
```

#### Create more complex models with Subschemas

> For more complex data mapping, the `x-adapter` library also provides a `subSchema` type that

Reviewer: I wouldn't say that `x-adapter` provides with `subSchemas`.  The idea of subschemas is being able to re-use a schema between different adapters and with different paths. Something like this maybe?. I would also change the example reusing the same schema under different paths in two adapters. 
> When you are creating adapters for different APIs you might find the case that you have to map the same model in different places. To help you with that, schemas allows to use subschemas. To use them you just have to provide with the path of the data to map, and the schema to apply to it.

receives a `Source`, a `Target` and a `Path` that will be applied to the inner path of the main
object. Furthermore, and not restricted to using subschemas, you can also use schema with function
return types, that makes schemas powerful to custom the logic you need in your data mapping.

```tsx
// Example of using a subSchema to map a response
const responseMapper = schemaMapperFactory<ApiResponseModel, SearchResponse>({
  totalResults: 'results.total',
  results: {
    $path: 'results.items',
    $subSchema: {
      id: 'id',
      name: 'name',
      url: 'result_url',
      images: ({ result }) => result.images.map(({ url }) => url),
      price: {
        value: ({ price, sale_price }) => (sale_price ? sale_price : price)
      }
    }
  }
});
```

#### Using a mutable schema

The `x-adapter` library also provides a `MutableSchema` type, a `Schema` that can replace an
original schema, override it or create a new one based on the original one. This can be useful to

> define a shared pattern that you will need to reuse in different endpoint adapters.

Reviewer: I wouldn't say that "a `MutableSchema` type, a `Schema` that can replace an
original schema, override it". 

> With the `x-adapter` you can create mutable schemas. These schemas are helpful when writing libraries where you want to have some defaults, but allow to modify or extend them.

I would also add an example for each one of the options that the mutable schemas has.

```tsx
import { createMutableSchema, Schema } from '@empathyco/x-adapter';
import { SearchRequest } from '@empathyco/x-types';
import { ApiRequestModel } from 'YOUR-API-SOURCE';

export const searchRequestSchema = createMutableSchema<Schema<SearchRequest, ApiRequestModel>>({
  q: 'query',
  limit: 'rows',
  offset: 'start'
  // ... any existing types from your API source
});
```

### Extend an adapter that uses schemas

You can check the
[x-platform-adapter](https://github.com/empathyco/x/tree/main/packages/x-adapter-platform) library.
You will find a sample implementation of the `x-adapter` library based on the
[Search Platform API](https://docs.empathy.co/develop-empathy-platform/api-reference/search-api.html),
and also some guidance on how to extend it for your needs.

> ## Test

Reviewer: I would add a little introduction, explaining that we use `jest` for the tests and where are they located.

The `x-adapter` library also provides a `MutableSchema` type, a `Schema` that can replace an
original schema, override it or create a new one based on the original one. This can be useful to
define a shared pattern that you will need to reuse in different endpoint adapters.

```tsx
import { createMutableSchema, Schema } from '@empathyco/x-adapter';
import { SearchRequest } from '@empathyco/x-types';
import { ApiRequestModel } from 'YOUR-API-SOURCE';

export const searchRequestSchema = createMutableSchema<Schema<SearchRequest, ApiRequestModel>>({
  q: 'query',
  limit: 'rows',
  offset: 'start'
  // ... any existing types from your API source
});
```

### Extend an adapter that uses schemas

You can check the
[x-platform-adapter](https://github.com/empathyco/x/tree/main/packages/x-adapter-platform) library.
You will find a sample implementation of the `x-adapter` library based on the
[Search Platform API](https://docs.empathy.co/develop-empathy-platform/api-reference/search-api.html),
and also some guidance on how to extend it for your needs.

## Test

```
npm test
```

## Changelog

[Changelog summary](https://github.com/empathyco/x/blob/main/packages/x-adapter/CHANGELOG.md)

> ## Contributors

Reviewer 2: I think I wouldn't add a specific section for contributors in the package, as we have one for the whole repository, and adding one to each package seems too granular and difficult to maintain.

| <a href="https://github.com/CachedaCodes"><img src="https://avatars.githubusercontent.com/u/7124620?v=4" width="100px;" alt=""/><br /><br /><b>Guillermo Cacheda</b></a> |      <a href="https://github.com/diegopf"><img src="https://avatars.githubusercontent.com/u/7504736?v=4" width="100px;" alt=""/><br /><br /><b>Diego Pascual</b></a>       | <a href="https://github.com/mnavarroespinosa"><img src="https://avatars.githubusercontent.com/u/77450928?v=4" width="100px;" alt=""/><br /><br /><b>Manuel Navarro</b></a> | <a href="https://github.com/ajperezbau"><img src="https://avatars.githubusercontent.com/u/75546736?v=4" width="100px;" alt=""/><br /><br /><b>Abraham Pérez</b></a> |

## Contributing

To start contributing to the project, please take a look to

> our **[Contributing Guide](https://github.com/empathyco/x/blob/HEAD/.github/CONTRIBUTING.md).**

Reviewer 2: I think is better use the link with the branch instead:

```suggestion
our **[Contributing Guide](https://github.com/empathyco/x/blob/main/.github/CONTRIBUTING.md).**
```

## License

> [empathyco/x License](https://github.com/empathyco/x/blob/main/LICENSE)

Reviewer 2: We should put the license file in every package. Could you add it to this package? You can copy the same file that we have in x-components package: https://github.com/empathyco/x/blob/main/packages/x-components/LICENSE

Then, in this README, link it to that new file.

        };
      }),
      total: total
    };
  }
});
```

<br>

#### Using a dynamic endpoint

You can use a mapper function to do so and pass the dynamic part of the url as a parameter when the
adapter function is called.

```ts
export const getProductById = endpointAdapterFactory({
  endpoint: ({ id }: GetProductByIdRequest) => 'https://dummyjson.com/products/' + id,
  responseMapper(response: Readonly<ApiProduct>): AppProduct {
    return productMap(response);
  }
});
```

Or you can place your parameter(s) inside curly brackets, and they will be added to the final
string, thanks to the
[interpolate](https://github.com/empathyco/x/blob/main/packages/x-adapter/src/utils/interpolate.ts)
utility function. If a value is not provided, its parameter would just will be removed from the
string.

```ts
export const getItemById = endpointAdapterFactory({
  endpoint: 'https://dummyjson.com/{section}/{id}',
  // ... rest of options to configure
  }
});

// You would pass the parameter's value in the function call
getItemById({ section: "products", id: 1 });
getItemById({ section: "quotes", id: 3 });
```

Additionally, you can also overwrite your adapter's endpoint definition using the

> `RequestOptions.endpoint` parameter in the function call. Take in account that your `responseMapper`

Reviewer 3: ```suggestion
`RequestOptions.endpoint` parameter in the function call. Take into account that your `responseMapper`
```

definition should be agnostic enough to support the change:

```ts
export const getItemById = endpointAdapterFactory({
  endpoint: 'https://dummyjson.com/quotes/{id}',
  // ... rest of options to configure

>   }

Reviewer 3: ```suggestion
```

[`options`](https://github.com/empathyco/x/blob/main/packages/x-adapter/src/http-clients/types.ts)
object to make the request with.

```ts
// HTTP Client
const customHttpClient: HttpClient = (endpoint, options) =>
  axios.get(endpoint, { params: options?.parameters }).then(response => response.data);

// Request Mapper
const customRequestMapper: Mapper<AppSearchRequest, ApiRequest> = ({ query }) => {
  return {
    q: query
  };
};

// Object Mapper to use in the Response Mapper
const productMap: Mapper<ApiProduct, AppProduct> = product => {
  return {
    id: product.id.toString(),
    name: product.title,
    price: product.price
  };
};

// Response Mapper
const customResponseMapper: Mapper<ApiSearchResponse, AppSearchResponse> = ({
  products,
  total
}) => {
  return {
    products: products.map(product => productMap(product, {})),
    total: total
  };
};

// Adapter factory function implementation
export const searchProducts = endpointAdapterFactory({
  endpoint: 'https://dummyjson.com/products/search',
  httpClient: customHttpClient,
  requestMapper: customRequestMapper,
  responseMapper: customResponseMapper
});
```

<br>

### Implement your own adapter using schemas

> The `x-adapter` library includes a `schemaMapperFactory` function that returns a mapper function to
> execute the `schema` you provide. The function accepts a `Source` object (the data to be
> transformed) and a `Target` object. You just need to create the request and response models to
> define the structure of your data vs the APIs data, and pass them to the factory function.

Reviewer: Let's try to justify using schemas.

```suggestion
Sometimes the transformations you will need to do in the mappers are just renaming parameters. What the API calls `q` might be called `query` in your request. To ease this transformations, `@empathyco/x-adapter` allows to create mappers using schemas.

A schema is just a dictionary where the key is the desired parameter name, and the value is the path of the source object that has the desired value or a simple mapper function if you need to transform the value somehow.
```

###### Types definition

```ts
// API data models
interface ApiUserRequest {
  q: string;
}
interface ApiUserResponse {
  users: ApiUser[];
  total: number;
}
interface ApiUser {
  id: number;
  firstName: string;
}

// App's data models
interface AppUserRequest {
  query: string;
}
interface AppUserResponse {
  people: AppUser[];
  total: number;
}
interface AppUser {
  id: string;
  name: string;
}
```

###### Schema's mapper factory function implementation

```ts

> // Object Mapper to use in the Response Mapper

Reviewer: Like above, I would just inline this mapper inside the `userSearchResponseMapper` `users.map(...)` call.

  people: {
    $subSchema: userSchema,
    $path: 'users'
  },
  count: 'total'
});

const requestMapper = schemaMapperFactory<SearchUsersRequest, ApiSearchUsersRequest>({
  q: 'query'
});

// Endpoint Adapter Factory function implementation
export const searchUsersWithContactInfo = endpointAdapterFactory({
  endpoint: 'https://dummyjson.com/users/search',
  requestMapper,
  responseMapper
});
```

<br>

#### Using a mutable schema

This feature lets you have some default mappers, and modify or extend them for some concrete
implementations. To do so, you should use the `createMutableSchema` helper function, and pass as a
parameter the schema you want to make mutable.

```ts
// TODO: Creating a mutable schema
```

Once you have your mutable schema, you can use its available methods to obtain a new schema based on
it:

- `replace`: Replaces completely the original Schema.
- `override`: Merges the original schema with the new one.
- `extends`: Creates a new Schema based on the original one. The original remains unchanged.

```ts
// TODO: Mutable schema's methods: '$replace', '$override', '$extends'
```

<br>

> ### Extend an adapter that uses schemas

Author: @javieri-empathy @diegopf: This is the part I doubt to add here this way -----> The idea I have is documenting it in `x-adapter-platform` and linking to that readme section when it's done instead of linking to the full package.

Reviewer: I would probably leave it here for the next PR. We can link the platform adapter as an example use case. Think that Extending an adapter is a feature provided by `x-adapter`. It is NOT an specific part of `x-platform-adapter`.

}
interface ApiProduct {
  id: number;
  title: string;
  price: number;
}

// App's data models
interface AppSearchRequest {
  query: string;
}
interface AppSearchResponse {
  products: AppProduct[];
  total: number;
}
interface AppProduct {
  id: string;
  name: string;
  price: number;
}
```

###### Adapter's factory function implementation

```ts
import { endpointAdapterFactory } from '@empathyco/x-adapter';

export const searchProducts = endpointAdapterFactory({
  endpoint: 'https://dummyjson.com/products/search',
  requestMapper({ query }: Readonly<AppSearchRequest>): ApiRequest {
    return {
      q: query // the request will be triggered as https://dummyjson.com/products/search?q=phone
    };
  },
  responseMapper({ products, total }: Readonly<ApiSearchResponse>): AppSearchResponse {
    return {
      products: products.map(product => {
        return {
          id: product.id.toString(),
          name: product.title,
          price: product.price
        };
      }),
      total: total
    };
  }
});

// Function call
async function searchOnClick() {
  const response = await searchProducts({ query: 'phone' });
  this.items = response.products;
}
```

<br>

> #### Using a dynamic endpoint

Reviewer: RE-posting because I believe this comment was lost when modifying the file 😅
I would probably invert this section. Start with the simple use case and then go to the advanced one.

Using a dynamic endpoint
If you need to generate an endpoint url dynamically, you can add parameters inside curly brackets to the endpoint string. By default these parameters will be replaced using the request data. If a parameter is not found inside the request, an empty string will be used.
```ts
export const getItemById = endpointAdapterFactory({
  endpoint: 'https://dummyjson.com/{section}/{id}',
  // ... rest of options to configure
  }
});
getItemById({ section: "products", id: 1 }); // 'https://dummyjson.com/products/1'
getItemById({ section: "quotes", id: 3 }); // 'https://dummyjson.com/quotes/3'
getItemById({ section: "quotes" }); // 'https://dummyjson.com/quotes/'
For more complex use cases, you can use a mapper function. This function receives the request, and must return the URL string.

export const getProductById = endpointAdapterFactory({
  endpoint: ({ id }: GetProductByIdRequest) => 'https://dummyjson.com/products/' + id,
  // ... rest of options to configure
});
```
